blockbuster/tob_lane.go

```diff
@@ -0,0 +1,340 @@
+		// if the transaction is already in the (partial) block proposal, we skip it
+		txHash, err := l.getTxHashStr(tmpBidTx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get bid tx hash: %w", err)
+		}
+		if _, ok := selectedTxs[txHash]; ok {
+			continue selectBidTxLoop
+		}
+
+		bidTxBz, err := l.prepareProposalVerifyTx(cacheCtx, tmpBidTx)
+		if err != nil {
+			txsToRemove[tmpBidTx] = struct{}{}
+			continue selectBidTxLoop
+		}
+
+		bidTxSize := int64(len(bidTxBz))
+		if bidTxSize <= maxTxBytes {
+			bidInfo, err := l.af.GetAuctionBidInfo(tmpBidTx)
+			if bidInfo == nil || err != nil {
+				// Some transactions in the bundle may be malformed or invalid, so we
+				// remove the bid transaction and try the next top bid.
+				txsToRemove[tmpBidTx] = struct{}{}
+				continue selectBidTxLoop
+			}
+
+			// store the bytes of each ref tx as sdk.Tx bytes in order to build a valid proposal
+			bundledTransactions := bidInfo.Transactions
+			sdkTxBytes := make([][]byte, len(bundledTransactions))
+
+			// Ensure that the bundled transactions are valid
+			for index, rawRefTx := range bundledTransactions {
+				refTx, err := l.af.WrapBundleTransaction(rawRefTx)
+				if err != nil {
+					// Malformed bundled transaction, so we remove the bid transaction
+					// and try the next top bid.
+					txsToRemove[tmpBidTx] = struct{}{}
+					continue selectBidTxLoop
+				}
+
+				txBz, err := l.prepareProposalVerifyTx(cacheCtx, refTx)
+				if err != nil {
+					// Invalid bundled transaction, so we remove the bid transaction
+					// and try the next top bid.
+					txsToRemove[tmpBidTx] = struct{}{}
+					continue selectBidTxLoop
+				}
+
+				sdkTxBytes[index] = txBz
+			}
+
+			// At this point, both the bid transaction itself and all the bundled
+			// transactions are valid. So we select the bid transaction along with
+			// all the bundled transactions. We also mark these transactions as seen and
+			// update the total size selected thus far.
+			tmpSelectedTxs = append(tmpSelectedTxs, bidTxBz)
+			tmpSelectedTxs = append(tmpSelectedTxs, sdkTxBytes...)
+
+			// Write the cache context to the original context when we know we have a
+			// valid top of block bundle.
+			write()
+
+			break selectBidTxLoop
+		}
+
+		txsToRemove[tmpBidTx] = struct{}{}
+		l.logger.Info(
+			"failed to select auction bid tx; tx size is too large",
+			"tx_size", bidTxSize,
+			"max_size", maxTxBytes,
+		)
+	}
+
+	// Remove all invalid transactions from the mempool.
+	for tx := range txsToRemove {
+		if err := l.Remove(tx); err != nil {
+			return nil, err
+		}
+	}
+
+	return tmpSelectedTxs, nil
+}
+
+// ProcessLane which verifies the lane's portion of a proposed block.
+func (l *TOBLane) ProcessLane(ctx sdk.Context, proposalTxs [][]byte) error {
+	for index, txBz := range proposalTxs {
+		tx, err := l.processProposalVerifyTx(ctx, txBz)
+		if err != nil {
+			return err
+		}
+
+		// skip transaction if it does not match this lane
+		if !l.Match(tx) {
+			continue
+		}
```

Comment:
Shouldn't `match` be called beforehand?

Reply:
It means we'll decode twice, but we save on verification. Why not. 

Addressed in 63d8f68